### PR TITLE
[CI] Pull requests migration to buildkite

### PR DESCRIPTION
**Pull request by pazone**
## What does this PR do?
Migrates pull request workflow from Jenkins and Github Actions to Buildkite

Migration of Jenkins jobs to new ephemeral VMs: 
- Ubuntu
- Windows
- macOS (orka)

Now the pipeline runs build and unit tests on all supported OS in buildkike.
Code coverage files are being downloaded by the `merge-coverage`. Due to platform dependent code we merge coverage results from all platforms and publish the merged coverage report to sonarQube. 

- Relates https://github.com/elastic/ingest-dev/issues/1710

### CODE REVIEW

**Review by dliappis — COMMENTED** [review]
Thank you for this massive work!
Basically LGTM from me, left some minor comments.

Is https://buildkite.com/elastic/elastic-agent/builds/4389 an example run where we can see this work in operation?

File: .buildkite/hooks/post-checkout.ps1 [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# Set error handling
+$ErrorActionPreference = "Stop"
+
+# Define a function to checkout and merge
+function Checkout-Merge {
```
[the remark this answers is not in the document]
Comment (pazone) [reply .buildkite/hooks/post-checkout.ps1:5]: That's cool! I'd like to try it in my next pull request. I need to land it today somehow :)

File: .buildkite/scripts/steps/merge.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+# Downloads and merges coverage files from multiple steps into a single file (build/TEST-go-unit.cov).
+# Usage: merge.sh <step1> <step2> ... Where <step> is the id of the step that contains the coverage artifact.#  
+
+set -exuo pipefail
```
[the remark this answers is not in the document]
Comment (pazone) [reply .buildkite/scripts/steps/merge.sh:6]: I think yes. I prefer the `-x` unless we use any secrets

**Review by michalpristas — APPROVED** [review]
from what i can tell it looks good. 
thanks for making upgrade more stable

File: internal/pkg/agent/application/upgrade/upgrade.go [hunk 3]
```diff
@@ -386,14 +386,16 @@ func copyDir(l *logger.Logger, from, to string, ignoreErrs bool) error {
 		}
 	}
 
+	// Try to detect if we are running with SSDs. If we are increase the copy concurrency,
+	// otherwise fall back to the default.
+	copyConcurrency := 1
```
Comment (cmacknz) [inline internal/pkg/agent/application/upgrade/upgrade.go:391]: I changed this a bit, if I merge https://github.com/elastic/elastic-agent/pull/3653 before you merge this you'll have a conflict to resolve.

File: .ci/Jenkinsfile [hunk 4]
```diff
@@ -95,57 +95,7 @@ pipeline {
...
-                }
-              }
-            }
-          }
-          stage('Test') {
-            steps {
-              withGithubNotify(context: "Test-${PLATFORM}") {
-                withMageEnv(){
-                  dir("${BASE_DIR}"){
-                    withEnv(["RACE_DETECTOR=true", "TEST_COVERAGE=${isCodeCoverageEnabled()}"]) {
```
Comment (blakerouse) [inline .ci/Jenkinsfile:117]: Looks like the `RACE_DETECTOR=true` is removed in the buildkite jobs. That needs to stay enabled.

**Review by blakerouse — CHANGES_REQUESTED** [review]
Overall the changes to make this run under buildkite look great. Only issue I see and the reason I am making it a blocker is because the RACE_DETECTOR was not enabled, something that we need to keep enabled when running the unit tests.

**Review by blakerouse — APPROVED** [review]
Thanks for enabling the race detector. Looks good now.

**Pull request merged**